### PR TITLE
Add a general request function to the client protocol.

### PR DESCRIPTION
**Pull request by adreyer**
There are a few places where we map request method keywords to
http-client function. The http client then maps them back to keywords
which is redundant. This adds a general request function to the protocol
that accepts a keyword method.

All the other methods could be tweaked to use request. I'm not sure if that's worth it.

### CODE REVIEW

**Comment by rlinehan** [comment]
@adreyer travis failures on this seem to be legitimate - I get the same when I run locally

```clojure
ERROR in (persistent-async-client-test) (AFn.java:429)
Uncaught exception, not in assertion.
expected: nil
  actual: clojure.lang.ArityException: Wrong number of args (2) passed to: async/eval12700/request-with-client--12705
                                                                  ...                        
puppetlabs.http.client.async/eval12724/create-client/fn/reify/request                         async.clj:  466
...
 ```

File: src/clj/puppetlabs/http/client/async.clj [hunk 1]
```diff
@@ -447,19 +447,21 @@
...
-      (trace [_ url opts] (request-with-client (assoc opts :method :trace :url url) nil client))
+      (trace [this url opts] (common/request this url :trace opts))
       (options [this url] (common/options this url {}))
-      (options [_ url opts] (request-with-client (assoc opts :method :options :url url) nil client))
+      (options [this url opts] (common/request this url :post opts))
       (patch [this url] (common/patch this url {}))
-      (patch [_ url opts] (request-with-client (assoc opts :method :patch :url url) nil client))
+      (patch [this url opts] (common/request this url :patch opts))
+      (request [this url method] (common/request this url method {}))
+      (request [_ url method opts] (request-with-client (assoc opts :method method :url url) client))
```
Comment (fpringvaldsen) [inline src/clj/puppetlabs/http/client/async.clj]: I think the test failures are coming from the fact that this call to `request-with-client` is missing an argument. The same thing is happening in the `request` function in the `sync` namespace.

**Comment by adreyer** [comment]
@fpringvaldsen @rlinehan this looks fixed was there anything else todo?

**Comment by rlinehan** [comment]
I think it would be good to have add some tests - probably to https://github.com/puppetlabs/clj-http-client/blob/master/test/puppetlabs/http/client/async_plaintext_test.clj#L74 and somewhere in https://github.com/puppetlabs/clj-http-client/blob/master/test/puppetlabs/http/client/sync_plaintext_test.clj.

**Comment by adreyer** [comment]
After using this I moved the new function to "make-request".

**Comment by adreyer** [comment]
I will soon be blocked on this @rlinehan @fpringvaldsen do either of you have time to review merge?

**Comment by cprice404** [comment]
@camlow325 I think @rlinehan is on PTO today so maybe you could review this since you're already in this repo

**Comment by fpringvaldsen** [comment]
The new tests look good and the tests are passing for me locally, so I'm going to go ahead and merge this.

**Pull request merged**